lib/sshutils/ctx.go

```diff
@@ -88,37 +125,44 @@ func (c *ConnectionContext) ExportEnv(m map[string]string) {
 }
 
 // GetAgent returns a agent.Agent which represents the capabilities of an SSH agent,
-// or nil if no agent is available in this context.
+// or nil if no agent is available in this context.  This is a legacy method used in
+// parts of Teleport which still expect persistent agent connections that are greedily
+// initialized as soon as forwarding capabilities are requested. Prefer using the
+// StartAgentChannel method directly where possible.
 func (c *ConnectionContext) GetAgent() agent.Agent {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-	return c.agent
+	if c.getForwardAgent() {
+		return teleagent.Lazy(c.Context, c.StartAgentChannel)
+	}
+	return nil
 }
 
-// GetAgentChannel returns the channel over which communication with the agent occurs,
-// or nil if no agent is available in this context.
-func (c *ConnectionContext) GetAgentChannel() ssh.Channel {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-	return c.agentChannel
+// SetForwardAgent configures this context to support agent forwarding.
+func (c *ConnectionContext) SetForwardAgent(forwardAgent bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.forwardAgent = forwardAgent
 }
 
-// SetAgent sets the agent and channel over which communication with the agent occurs.
-func (c *ConnectionContext) SetAgent(a agent.Agent, channel ssh.Channel) {
+// getForwardAgent loads the forwardAgent flag with lock.
+func (c *ConnectionContext) getForwardAgent() bool {
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	if c.agentChannel != nil {
-		c.agentChannel.Close()
-	}
-	c.agentChannel = channel
-	c.agent = a
+	return c.forwardAgent
 }
 
 // AddCloser adds any closer in ctx that will be called
 // when the underlying connection is closed.
 func (c *ConnectionContext) AddCloser(closer io.Closer) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
+	// if context was already closed, run the closer immediately
+	// in the background.
+	if c.closed {
+		go func() {
+			closer.Close()
+		}()
```

Comment:
this can be: `go closer.Close()`
but why in a goroutine?

Reply:
Calling `AddCloser` in a scenario where `ConnectionContext` has already started closing should be a fairly rare event.  As such, I figure its better from a debugging perspective to be able to assume that closers are always run asynchronously, rather than having a small chance of the method blocking on the closer.